src/pages/index.js

```diff
@@ -80,64 +74,185 @@ const addCardPopup = new PopupWithForm(
 );
 addCardPopup.setEventListeners();
 
+
+//  
+// Image Preview Popup
+//  
+
 const popupImageModal = new PopupWithImage('#preview-image-modal');
 popupImageModal.setEventListeners();
 
-const cardSection = new Section(
-	{
-		items: initialCards,
-		renderer: renderCard,
-	},
-	'.cards__list'
-);
-cardSection.renderItems();
-
-// function openPopup(modal) {
-// 	modal.classList.add('modal_opened');
-// 	document.addEventListener('keydown', escPopup);
-// }
-
-// function closePopup(modal) {
-// 	modal.classList.remove('modal_opened');
-// 	document.removeEventListener('keydown', escPopup);
-// }
-
-// function escPopup(e) {
-// 	if (e.key === 'Escape') {
-// 		const modal = document.querySelector('.modal_opened');
-// 		if (modal) closePopup(modal);
-// 	}
-// }
-
-function renderCard(cardData) {
-	const card = new Card(cardData, '#card-template', handleImageClick);
-	const cardElement = card.getView();
-	cardSection.addItem(cardElement);
+function handleImageClick(name, link) {
+	popupImageModal.open({ name, link });
 }
 
-function handleOverlayClose(e) {
-	if (e.target.classList.contains('modal_opened')) {
-		popupImageModal.close();
-	}
+//  
+// Form Validation
+//  
+
+const addEditValidator = new FormValidator(config, editFormElement);
+addEditValidator.enableValidation();
+
+const addCardValidator = new FormValidator(config, addFormElement);
+addCardValidator.enableValidation();
+
+const updateAvatarValidator = new FormValidator(config, avatarFormElement);
+updateAvatarValidator.enableValidation();
+
+//  
+// User Info
+//  
+
+const userInfo = new UserInfo({
+	nameSelector: '.profile__title',
+	jobSelector: '.profile__description',
+	userAvatar: '.profile__image',
+});
+
+// 
+// Functions
+//  
+
+function createCard(cardData) {
+	const card = new Card(
+		cardData,
+		'#card-template',
+		handleImageClick,
+		handleTrashClick,
+		handleLikeClick,
+	);
+	cardSection.addItem(card.generateCard());
 }
 
-function handleImageClick(name, link) {
-	popupImageModal.open(name, link);
+// ADD CARD AND EDIT PROFILE
+
+function handleAddCardFormSubmit(data) {
+	addCardPopup.submitButtonState(true, "Saving...");
+	api
+		.addCard(data)
+		.then((data) => {
+			createCard(data);
+		})
+		.then(() => {
+			addCardPopup.close();
+		})
+		.catch((err) => {
+			console.log(err);
+		})
+		.finally(() => {
+			addCardPopup.submitButtonState(false);
+		});
 }
 
-function handleProfileEditSubmit(value) {
-	userInfo.setUserInfo(value);
-	editProfilePopup.close();
+function handleProfileEditSubmit(data) {
+	editProfilePopup.submitButtonState(true, "Saving...");
+	api
+	.updateEditProfile(data)
+		.then((data) => {
+			userInfo.setUserInfo(data);
+			editProfilePopup.close();
+		})
+		.catch((err) => {
+			console.log(err);
+		})
+		.finally(() => {
+			editProfilePopup.submitButtonState(false);
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```